### DISCUSSION TITLE
Support OA3 Server Templates

### DISCUSSION THREAD

**Discussion by DaleyKD** [Burgyn/MMLib.SwaggerForOcelot#155] [Ideas]
https://swagger.io/docs/specification/api-host-and-base-path/

I'm trying to serve my swagger.json from two locations:
* https://acme.company.com/appName/myTenant/swagger/v1/swagger.json [_this is a direct URL to the .json, which lives "under" the application I've written_]
* https://gateway.company.com/swagger [_the MMLib Swagger UI for Ocelot_]

In an attempt to support this, I started using the `servers` part of the OpenAPI 3 spec.  It looks something like this in C#:
```csharp
new OpenApiServer()
{
    Url = "https://{accountHost}.company.com/appName/{tenantHost}",
    Description = "My Application Direct",
    Variables = new Dictionary<string, OpenApiServerVariable>()
    {
        {
            "accountHost", new OpenApiServerVariable()
            {
                Description = "The hostname of the account.",
                Default = "acme"
            }
        },
        {
            "tenantHost", new OpenApiServerVariable()
            {
                Description = "The hostname of the tenant.",
                Default = "myTenant"
            }
        }
    }
}
```

Reading in MMLib, I noticed you only support one server for now.  That's totally understandable.  To get this to semi-work, I then added the server at the 0-index:
```csharp
new OpenApiServer() {Url = "/{tenantHost}", Description = "My Application via Gateway"}
```

However, to be a bit more verbose, I'd prefer to use:
```csharp
new OpenApiServer()
{
    Url = "https://gateway.company.{tld}/{tenantHost}",
    Description = "My Application via Gateway",
    Variables = new Dictionary<string, OpenApiServerVariable>()
    {
            "tld", new OpenApiServerVariable()
            {
                Description = "The environment.",
                Default = "com",
                Enum = new List<string>()
                {
                    "com",
                    "net",
                    "co.uk"
                }
            }
        {
            "tenantHost", new OpenApiServerVariable()
            {
                Description = "The hostname of the tenant.",
                Default = "myTenant"
            }
        }
    }
}
```

This throws a `UriFormatException`.  It seems like https://github.com/Burgyn/MMLib.SwaggerForOcelot/blob/18ea188720e6a0110ec7c3ef340a7a884d4c4121/src/MMLib.SwaggerForOcelot/Transformation/SwaggerJsonTransformer.cs#L78
is not designed to support server templates.

**Comment by Burgyn** [comment]
Hi @DaleyKD,

I'm sorry for late answer.

I will try to explore the possibilities and try to respond within a few days.

**Reply by DaleyKD** [reply]
Appreciate the consideration!

**Reply by Burgyn** [reply]
Hi @DaleyKD,

I prepared a new version `4.1.0` with the ability to take over the definition of OpenAPI servers from the downstream service.

Please test if this solves your use case.

---
from docs:

If you have multiple servers defined in the downstream service Open API documentation, or you use server templating and you want to use it on the gateway side as well, then you must explicitly enable it on the Swagger endpoint definition by setting property `TakeServersFromDownstreamService` to `true`.

```json
"SwaggerEndPoints": [
    {
      "Key": "users",
      "TakeServersFromDownstreamService": true,
      "Config": [
        {
          "Name": "Users API",
          "Version": "v1",
          "Service": {
            "Name": "users",
            "Path": "/swagger/v1/swagger.json"
          }
        }
      ]
    }
]
```

> ⚠ If you set `TakeServersFromDownstreamService` to `true`, then the server path is not used to transform the paths of individual endpoints.
